modules/custom/openy_programs_search/src/DataStorage.php

```diff
@@ -0,0 +1,905 @@
+    return $data;
+  }
+
+  /**
+   * Get sessions by Program ID & Location ID.
+   *
+   * @param int $program_id
+   *   Program ID.
+   * @param int $location_id
+   *   Location ID.
+   *
+   * @return array
+   *   List of sessions.
+   */
+  public function getSessionsByProgramAndLocation($program_id, $location_id) {
+    $cid = __METHOD__ . $program_id . $location_id;
+    if ($cache = $this->cache->get($cid)) {
+      return $cache->data;
+    }
+
+    $params = [
+      'program' => $program_id,
+      'branch' => $location_id,
+    ];
+
+    $data = $this->client->getSessions($params);
+
+    $this->cache->set($cid, $data);
+    return $data;
+  }
+
+  /**
+   * Get registration link.
+   *
+   * @param int $program_id
+   *   Program ID.
+   * @param int $session_id
+   *   Session ID.
+   *
+   * @return string
+   *   Registration link.
+   */
+  public function getRegistrationLink($program_id, $session_id) {
+    $uri = $this->getUrlFromOpenyProgramsSearchSettings('registration_path');
+
+    $query = [
+      'program_id' => $program_id,
+      'session_ids' => $session_id,
+    ];
+
+    $path = Url::fromUri(
+      $uri,
+      [
+        'query' => $query,
+        'absolute' => TRUE,
+        'https' => TRUE,
+      ]
+    );
+
+    return $path->toString();
+  }
+
+  /**
+   * Get child registration link.
+   *
+   * @param int $school_id
+   *   School ID.
+   * @param int $program_id
+   *   Program ID.
+   * @param int $context_id
+   *   Rate option ID.
+   *
+   * @return string
+   *   Registration link.
+   */
+  public function getChildCareRegistrationLink($school_id, $program_id, $context_id) {
+    $domain = $this->getUrlFromOpenyProgramsSearchSettings('base_url');
+
+    $map = $this->getMapRateOptions();
+    $item = $map[$school_id][$program_id][$context_id];
+    return $domain . $item['registration_url'];
+  }
+
+  /**
+   * Get Location IDs by Childcare Program ID.
+   *
+   * @param int $program_id
+   *   Program ID.
+   *
+   * @return array
+   *   List of location IDs.
+   *
+   * @ingroup legacy
+   */
+  public function getLocationsByChildCareProgramId($program_id) {
+    $programMap = [];
+
+    $cid = __METHOD__ . $program_id;
+    if ($cache = $this->cache->get($cid)) {
+      $programMap = $cache->data;
+    }
+    else {
+      $locations = $this->getDaxkoLocationMap();
+      $locsr = array_flip($locations);
+      $locexp = $this->customLocationStem($locations);
+
+      $programs = $this->getAllChildCarePrograms();
+      foreach ($programs as $program) {
+        $loc = strstr($program->name, ' ', TRUE);
+
+        if (isset($locsr[$loc])) {
+          $programMap[$program->id][$locsr[$loc]] = $locsr[$loc];
+        }
+        else {
+          $ranking = [];
+          foreach ($locexp as $k => $exps) {
+
+            $ranking[$k] = 0;
+            if (substr($program->name, 0, 1) == substr($exps[0], 0, 1)) {
+              if ((strpos($program->name, "South Lake Houston") !== FALSE)) {
```

Comment:
Better code comments would help :( All I know is 'South Lake Houston' was being given -10 less weight. I'm guessing the functionality could apply differently for each site, maybe needing multiple programs to have different weight adjustments. But this just subtracts 10 it doesn't pin the item to the top, it's a bit confusing.